src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -278,21 +347,66 @@ private void UpdateDoubleCustomTags(double value, ReadOnlySpan<KeyValuePair<stri
         {
             try
             {
-                var index = this.FindMetricAggregatorsCustomTag(tags);
-                if (index < 0)
+                int index;
+                do
                 {
-                    // TODO: Measurement dropped due to MemoryPoint cap hit.
-                    return;
+                    index = this.FindMetricAggregatorsCustomTag(tags);
+                    if (index < 0)
+                    {
+                        // TODO: Measurement dropped due to MemoryPoint cap hit.
+                        return;
+                    }
                 }
-
-                this.metricPoints[index].Update(value);
+                while (!this.TryUpdateMetricPoint(index, value));
             }
             catch (Exception)
             {
                 // TODO: Measurement dropped due to internal exception.
             }
         }
 
+        private bool TryUpdateMetricPoint(int index, long value)
+        {
+            ref var metricPoint = ref this.metricPoints[index];
+            var metricPointLock = this.metricPointLocks[index];
+
+            metricPointLock.EnterReadLock();
+
+            if (metricPoint.MetricPointStatus == MetricPointStatus.Unset)
+            {
+                metricPointLock.ExitReadLock();
+                return false;
+            }
+
+            metricPoint.Update(value);
```

Comment:
haven't fully analyzed the implications, but I think we should move the status management to be entirely within the MetricPoint class, and avoid any locks in `AggregatorStore`.

1. MetricPoint already has `lock` for Histogrampoints now. So it can safely do Update() and SnapShot() adjusting the MetricPointStatus accordingly.
2. For SUM, there is no lock. But the following could achieve what we need, without the need of AggregatorStore entering ReadLocks

Update:
Do the update as before, and **after** that, always set status to `CollectPending`.
SnapShot:
Read the value as before. 

If the value has not changed since it read, then set status to `NoPendingUpdate`. (i.e no other sync instrument.update has occurred). If an update ever comes before next SnapShot, the next snapshot will include it. If no update, then cleanup time.

Else (i.e some other sync instrument update happened since snapshot did the "snapshotting(interlocked.read)"), do not modify the status. This means the update which just came, will be collected in the next cycle..

Reply:
to avoid the issue where an update gets lost when a CandidateForRemoval is cleanedup, we might need to do a lock in Update() if the previous status is CandidateForRemoval.

Reply:
Just an update, I haven't pushed it up yet, but I'm pretty close to implementing this feedback.

I've moved the status management down into the MetricPoint class itself, and I agree it's cleaner. I'm just carefully reviewing the status management before pushing it up.

I'll address your other feedback on this PR once I'm comfortable with things here. I think I've successfully simplified a few things like removing the `metricPointFreeList` as it may no longer be necessary. I've also reduced the set of `MetricPointStatus` enumeration values.

Reply:
yea. the other part about re-using histogram arrary is definitely separate. (As i was thinking about it, I think we might have more optimizations possible there for string[] object[] as well, when View is configured) Will chat about it after the cleanup part is stable.